src/ResourceHelperTrait.php

```diff
@@ -0,0 +1,92 @@
+<?php
+/*
+ * Copyright 2022 Google LLC
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are
+ * met:
+ *
+ *     * Redistributions of source code must retain the above copyright
+ * notice, this list of conditions and the following disclaimer.
+ *     * Redistributions in binary form must reproduce the above
+ * copyright notice, this list of conditions and the following disclaimer
+ * in the documentation and/or other materials provided with the
+ * distribution.
+ *     * Neither the name of Google Inc. nor the names of its
+ * contributors may be used to endorse or promote products derived from
+ * this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
+ * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+ * OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+ * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+
+namespace Google\ApiCore;
+
+use Google\ApiCore\ValidationException;
+
+trait ResourceHelperTrait
+{
+    private static $templateMap;
+
+    // Must be implemented by extendees to call loadTemplates.
+    private static function registerTemplates()
```

Comment:
Is the point of this method to allow the GapicClients to lazy load the templates? Currently, `GapicClientTrait` loads teh descriptors in `setClientOptions`, which is called from the `__constructor`, so we could just load all the templates up front and then we wouldn't need this method.

Reply:
Then, we could potentially be smarter and make it use values in `$options`, so we don't have to define `registerTemplates` on every GAPIC client:

```php
private static function registerTemplates()
{
        if (!is_null(self::$templateMap)) {
            return;
        }
        if (empty($this->options['descriptorsConfigPath']) || empty($this->options['serviceName'])) {
            return;
        }
        $descriptors = require($this->options['descriptorsConfigPath']);
        $templates = $descriptors['interfaces'][$this->options['serviceName']]['templateMap'] ?? [];
        self::$templateMap = [];
        foreach ($templates as $name => $template) {
             self::$templateMap[$name] = new PathTemplate($template);
        }
}
```

Reply:
Sorry, I'm confused, we can't use `$this` in a `static` method, can we? There is no instance associated with a `static` method.

Reply:
You're right... we can't. I missed that it was static. The way we have it now, `descriptorsConfigPath` and `serviceName` are both configurable in the constructor. So if those values were configured manually, the paths defined `registerTemplates` would differ.

Reply:
e.g.

```php
$foo = new FooClient(['descriptorsConfigPath' => '/path/to/my_descriptors.php', 'serviceName' => 'myServiceName']);
$templateName = $foo->templateName(); // This will still use the default descriptors and service name
```

Reply:
The way we currently have it isn't dynamic either, so we could still implement it this way, and it wouldn't be breaking. It's just a bit strange since the API makes it seem configurable and it wouldn't be.

Reply:
Yeah this is a fair point, but fwiw that's not a use case that we even support today for the existing resource helper support. It could be a cool thing to support in the future though!

I don't see why `serviceName` would ever change for a given GAPIC - it is generated based on a specific service. But I could definitely see a world where some veneer wants to extend the GAPIC and add more resource name mappings that aren't in the protos for whatever reason.

Reply:
I could even see it being used as a backwards compatibility shim if a pattern or type changes. We could encode the broken pattern/mapping in the veneer layer and extend the descriptor config.

Reply:
I think I liked it mostly because it opens us up to support a more config driven approach vs. hardcoded mapping like before. @dwsupplee gets credit for the idea.

Reply:
Resolving for now.